File: problems/020-offline-dependency-bundle/solutions/python/main.py
import sys

it = iter(sys.stdin.read().split())
n = int(next(it))
m = int(next(it))
lock = []
total = 0
for _ in range(n):
    next(it)
    d = next(it)
    z = int(next(it))
    lock.append((d, z))
    total += z
pay = [(next(it), int(next(it))) for _ in range(m)]
lock.sort()
pay.sort()
req = []
i = 0
while i < n:
    j = i + 1
    while j < n and lock[j][0] == lock[i][0]:
        j += 1
    if any(lock[k][1] != lock[i][1] for k in range(i + 1, j)):
        print("LOCK_CONFLICT", lock[i][0])
        raise SystemExit
    req.append(lock[i])
    i = j
for i in range(1, m):
    if pay[i][0] == pay[i - 1][0]:
        print("DUPLICATE_PAYLOAD", pay[i][0])
        raise SystemExit
missing = extra = size_error = None
i = j = 0
while i < len(req) or j < len(pay):
    if j == len(pay) or (i < len(req) and req[i][0] < pay[j][0]):
        if missing is None:
            missing = req[i][0]
        i += 1
    elif i == len(req) or pay[j][0] < req[i][0]:
        if extra is None:
            extra = pay[j][0]
        j += 1
    else:
        if req[i][1] != pay[j][1] and size_error is None:
            size_error = req[i][0]
        i += 1
        j += 1
if missing is not None:
    print("MISSING", missing)
    raise SystemExit
if extra is not None:
    print("EXTRA", extra)
    raise SystemExit
if size_error is not None:
    print("SIZE", size_error)
    raise SystemExit
unique = sum(z for _, z in req)
print("VALID", len(req), unique, total - unique)

File: problems/020-offline-dependency-bundle/solutions/javascript/main.js
import * as std from "std";
/** @typedef {[string, bigint]} R */
const a = std.in.readAsString().trim().split(/\s+/);
let p = 0,
  n = Number(a[p++]),
  m = Number(a[p++]);
/** @type {R[]} */
const lock = [];
/** @type {R[]} */
const pay = [];
let total = 0n;
for (let i = 0; i < n; i++) {
  p++;
  const d = a[p++], z = BigInt(a[p++]);
  lock.push([d, z]);
  total += z;
}
for (let i = 0; i < m; i++) pay.push([a[p++], BigInt(a[p++])]);
lock.sort((x, y) => x[0] < y[0] ? -1 : x[0] > y[0] ? 1 : 0);
pay.sort((x, y) => x[0] < y[0] ? -1 : x[0] > y[0] ? 1 : 0);
/** @type {R[]} */
const req = [];
/** @type {string | null} */
let out = null;
for (let i = 0; i < n;) {
  let j = i + 1;
  while (j < n && lock[j][0] === lock[i][0]) {
    if (lock[j][1] !== lock[i][1]) out = `LOCK_CONFLICT ${lock[i][0]}`;
    j++;
  }
  req.push(lock[i]);
  if (out) break;
  i = j;
}
if (!out) {
  for (let i = 1; i < m; i++) {
    if (pay[i][0] === pay[i - 1][0]) {
      out = `DUPLICATE_PAYLOAD ${pay[i][0]}`;
      break;
    }
  }
}
let unique = 0n;
if (!out) {
  let i = 0, j = 0;
  /** @type {string | null} */
  let missing = null;
  /** @type {string | null} */
  let extra = null;
  /** @type {string | null} */
  let sizeError = null;
  while (i < req.length || j < pay.length) {
    if (j === pay.length || i < req.length && req[i][0] < pay[j][0]) {
      if (missing === null) missing = req[i][0];
      i++;
    } else if (i === req.length || pay[j][0] < req[i][0]) {
      if (extra === null) extra = pay[j][0];
      j++;
    } else {
      if (req[i][1] !== pay[j][1] && sizeError === null) sizeError = req[i][0];
      i++;
      j++;
    }
  }
  if (missing !== null) out = `MISSING ${missing}`;
  else if (extra !== null) out = `EXTRA ${extra}`;
  else if (sizeError !== null) out = `SIZE ${sizeError}`;
  else for (const [, z] of req) unique += z;
}
if (!out) out = `VALID ${req.length} ${unique} ${total - unique}`;
std.out.puts(out + "\n");
